enhancements/microshift/lvms-topolvm-downstream-replacement.md

```diff
@@ -0,0 +1,269 @@
+   * Errors of the LVMCluster deployment will be written to the lvms-controller.
+
+### API Extensions
+
+MicroShift will now introduce the `LVMCluster`, `LVMVolumeGroup` and `LVMVolumeGroupNodeStatus` CustomResources.
+All existing documentation of LVMS on the CR usages applies and can be cross-referenced if users
+want a simplified setup of their storage devices.
+
+### Topology Considerations
+
+#### Hypershift / Hosted Control Planes
+
+N/A
+
+#### Standalone Clusters
+
+N/A
+
+#### Single-node Deployments or MicroShift
+
+This enhancement only applies to MicroShift.
+
+### Implementation Details/Notes/Constraints
+
+#### Enablement of configuration of no deviceClasses in LVMS
+
+Currently, LVMS does not allow for the configuration of no deviceClasses in the LVMCluster.
+This is a constraint that we will have to work around by allowing the user to specify
+an empty deviceClass in the LVMCluster and then coordinating that with the vg-manager.
+We have a prototype implementation that can be pushed into GA in time for the 4.17 release based on
+[OCPEDGE-830](https://issues.redhat.com//browse/OCPEDGE-830) that proves this is possible.
+
+#### Switching from TopoLVM to LVMS
+
+The switch from TopoLVM to LVMS will be done in a single release. We will support
+upgrading from 4.16 to 4.16, and from 4.14 to 4.16 (EUS).
+
+The switch will be done by removing the TopoLVM deployment and installing the LVMS
+deployment. The LVMS deployment will be configured to use the existing deviceClasses
+and storageClasses. There will be leftover images from the TopoLVM deployment that
+can be removed manually from the registry after the upgrade.
+
+There should not be other regressions in the upgrade process, as the LVMS migration
+is transparent to users.
+
+### Risks and Mitigations
+
+There is some risk that the LVMS deployment will not work with the existing
+deviceClasses and storageClasses. We will mitigate this risk by testing the
+upgrade process in CI and by providing a rollback mechanism in case of failure.
+Specific test scenarios that are not handled in the generic LVMS use cases should especially involve
+[Software RAIDs introduced with LVM Raid](https://github.com/topolvm/topolvm/blob/2597dc75821009230af3be4067d377eef4bbe0a4/docs/lvmd.md?plain=1#L28-L36) with custom create options
+This is because in LVMS, these are not supported by default and require a custom deviceClass
+and are now enabled for the first time within LVMS through the MicroShift deployment.
+
+Additionally, we will provide documentation on how to correctly set up an empty
+LVMCluster and how to run debugging procedures in case of failure.
+Any issues discovered during this process will be remediated with custom adjustments.
+
+### Drawbacks
+
+The main draw back of this change is that it might require manual remediation
+if a failure occurs during the upgrade process. Since this is a one-time change,
+we believe that the benefits of moving to LVMS outweigh this potential issue.
```

Comment:
A potential for manual intervention is a big deal on edge devices because it may actually brick the remote devices. Can you explain more about what kind of manual work might be required in case of rollback and whether it can be done without gaining physical access to the device?

Reply:
This should be covered with a usual rollback scenario, so I am assuming a rollback test should work fine?

Reply:
Yes, we have such tests and we will need to make sure it tests LVMS upgrade+downgrade.

Going back to this proposal, I think we need to soften the `manual remediation` statement. Instead, we should mentioned greenboot, rollback, etc.

Reply:
Reworded this to include Upgrade/Downgrade behavior testing via greenboot on top of making sure that we can revert the change by rolling back the assets atomically.